### PR TITLE
OTT-1381: Moving Vast unwrap module into Openwrap module

### PR DESCRIPTION
**Pull request by pm-jaydeep-mohite**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: modules/pubmatic/openwrap/openwrap_sshb_test.go [hunk 1]
```diff
@@ -8,12 +8,12 @@ import (
 	"github.com/magiconair/properties/assert"
 	mock_cache "github.com/prebid/prebid-server/modules/pubmatic/openwrap/cache/mock"
 	"github.com/prebid/prebid-server/modules/pubmatic/openwrap/models"
-	vastmodels "github.com/prebid/prebid-server/modules/pubmatic/vastunwrap/models"
+	//vastmodels "github.com/prebid/prebid-server/modules/pubmatic/vastunwrap/models"
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/openwrap_sshb_test.go:11]: remove this line

File: modules/pubmatic/openwrap/hook_raw_bidder_response.go [hunk 2]
```diff
@@ -1,42 +1,31 @@
...
 )
 
-func (m VastUnwrapModule) handleRawBidderResponseHook(
+func (m OpenWrap) handleRawBidderResponseHook(
 	miCtx hookstage.ModuleInvocationContext,
 	payload hookstage.RawBidderResponsePayload,
 	unwrapURL string,
 ) (result hookstage.HookResult[hookstage.RawBidderResponsePayload], err error) {
-	vastRequestContext, ok := miCtx.ModuleContext[RequestContext].(models.RequestCtx)
+	vastRequestContext, ok := miCtx.ModuleContext["rctx"].(models.RequestCtx)
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/hook_raw_bidder_response.go:18]: nitpick: use `models.RequestContext`

File: modules/pubmatic/openwrap/hook_raw_bidder_response.go [hunk 3]
```diff
@@ -1,42 +1,31 @@
...
-			fmt.Sprintf("found request without sshb=1 in handleRawBidderResponseHook() for pubid:[%d]", vastRequestContext.PubID))
-	}
 
-	vastRequestContext.VastUnwrapEnabled = vastUnwrapEnabled
 	if vastRequestContext.VastUnwrapEnabled {
 		// Do Unwrap and Update Adm
 		wg := new(sync.WaitGroup)
 		for _, bid := range payload.Bids {
-			if string(bid.BidType) == MediaTypeVideo {
+			if string(bid.BidType) == "video" {
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/hook_raw_bidder_response.go:28]: nitpick: use `models.MediaTypeVideo`

File: modules/pubmatic/openwrap/hook_raw_bidder_response.go [hunk 4]
```diff
@@ -49,11 +38,11 @@ func (m VastUnwrapModule) handleRawBidderResponseHook(
 		changeSet.RawBidderResponse().Bids().Update(payload.Bids)
 		result.ChangeSet = changeSet
 	} else {
-		vastRequestContext.VastUnwrapStatsEnabled = openwrap.GetRandomNumberIn1To100() <= m.StatTrafficPercentage
+		vastRequestContext.VastUnwrapStatsEnabled = GetRandomNumberIn1To100() <= m.cfg.Features.VASTUnwrapStatsPecent
 		if vastRequestContext.VastUnwrapStatsEnabled {
 			// Do Unwrap and Collect stats only
 			for _, bid := range payload.Bids {
-				if string(bid.BidType) == MediaTypeVideo {
+				if string(bid.BidType) == "video" {
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/hook_raw_bidder_response.go:45]: nitpick: use models.MediaTypeVideo

File: modules/pubmatic/openwrap/metrics/prometheus/prometheus.go [hunk 5]
```diff
@@ -250,6 +256,22 @@ func newMetrics(cfg *config.PrometheusMetrics, promRegistry *prometheus.Registry
...
 		[]string{pubIDLabel, profileIDLabel},
 	)
+	metrics.requests = newCounter(cfg, promRegistry,
+		"vastunwrap_status",
+		"Count of vast unwrap requests labeled by status",
+		[]string{pubIdLabel, bidderLabel, statusLabel})
+	metrics.wrapperCount = newCounter(cfg, promRegistry,
+		"vastunwrap_wrapper_count",
+		"Count of vast unwrap levels labeled by bidder",
+		[]string{pubIdLabel, bidderLabel, wrapperCountLabel})
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/metrics/prometheus/prometheus.go:266]: count(`vastunwrap_status`) should give total request value. Is separate `vastunwrap_wrapper_count` count needed?.

File: modules/pubmatic/openwrap/unwrap_service.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
+package openwrap
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/unwrap_service.go:1]: would be more clean to keep unwrap code under vastunwrap package

File: modules/pubmatic/openwrap/openwrap_sshb.go [hunk 7]
```diff
@@ -53,7 +52,7 @@ func (ow *OpenWrap) GetFeature() publisherfeature.Feature {
 }
 
 // GetVastUnwrapEnabled return whether to enable vastunwrap or not
-func GetVastUnwrapEnabled(rctx vastmodels.RequestCtx, VASTUnwrapTraffic int) bool {
+func GetVastUnwrapEnabled(rctx models.RequestCtx, VASTUnwrapTraffic int) bool {
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/openwrap_sshb.go:55]: This function is not needed

File: modules/pubmatic/openwrap/abtest.go [hunk 8]
```diff
@@ -12,6 +12,17 @@ func CheckABTestEnabled(rctx models.RequestCtx) bool {
...
+// based on random number generation
+func DisableVASTUnwrapConfigForRequest(partnerConfig map[int]map[string]string) map[int]map[string]string {
+	//create copy of the map
+	newPartnerConfig := copyPartnerConfigMap(partnerConfig)
+
+	// Disable enableVastUnwrapper key
+	newPartnerConfig[-1][models.VastUnwrapperEnableKey] = "0"
+	return newPartnerConfig
+}
+
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/abtest.go:25]: Removed copying of entire map for single update.

File: modules/pubmatic/openwrap/beforevalidationhook.go [hunk 9]
```diff
@@ -153,6 +153,18 @@ func (m OpenWrap) handleBeforeValidationHook(
...
 
+	// Re-use AB Test flag for VAST unwrap feature
+	if models.GetVersionLevelPropertyFromPartnerConfig(rCtx.PartnerConfigMap, models.VastUnwrapperEnableKey) == "1" {
+		randomNumber := GetRandomNumberIn1To100()
+		if randomNumber <= m.cfg.Features.VASTUnwrapPecent {
+			rCtx.ABTestConfigApplied = 1
+			rCtx.VastUnwrapEnabled = true
+		} else {
+			rCtx.PartnerConfigMap = DisableVASTUnwrapConfigForRequest(rCtx.PartnerConfigMap)
+			result.Warnings = append(result.Warnings, "update the rCtx.PartnerConfigMap for VAST Disable")
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/beforevalidationhook.go:164]: not required, removed

File: modules/pubmatic/openwrap/beforevalidationhook.go [hunk 10]
```diff
@@ -153,6 +153,18 @@ func (m OpenWrap) handleBeforeValidationHook(
 		result.Warnings = append(result.Warnings, "update the rCtx.PartnerConfigMap with ABTest data")
 	}
 
+	// Re-use AB Test flag for VAST unwrap feature
+	if models.GetVersionLevelPropertyFromPartnerConfig(rCtx.PartnerConfigMap, models.VastUnwrapperEnableKey) == "1" {
+		randomNumber := GetRandomNumberIn1To100()
+		if randomNumber <= m.cfg.Features.VASTUnwrapPecent {
+			rCtx.ABTestConfigApplied = 1
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/beforevalidationhook.go:160]: yes, we need it 
We would be enabling either VAST unwrap or AB test for given profile

File: modules/pubmatic/openwrap/entrypointhook.go [hunk 11]
```diff
@@ -136,6 +137,11 @@ func (m OpenWrap) handleEntrypointHook(
 	return result, nil
 }
 
+func getVastUnwrapperEnable(ctx context.Context, field string) bool {
+	vastEnableUnwrapper, _ := ctx.Value(field).(string)
+	return vastEnableUnwrapper == "1"
+}
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/entrypointhook.go:143]: Moved to openwrap_sshb.go

File: modules/pubmatic/openwrap/openwrap_sshb_test.go (hunk 1 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/openwrap_sshb_test.go:11]: Done

File: modules/pubmatic/openwrap/hook_raw_bidder_response.go (hunk 2 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/hook_raw_bidder_response.go:18]: Done

File: modules/pubmatic/openwrap/hook_raw_bidder_response.go (hunk 3 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/hook_raw_bidder_response.go:28]: Done

File: modules/pubmatic/openwrap/hook_raw_bidder_response.go (hunk 4 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/hook_raw_bidder_response.go:45]: Done

File: modules/pubmatic/openwrap/openwrap_sshb.go (hunk 7 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/openwrap_sshb.go:55]: removed

File: modules/pubmatic/openwrap/metrics/prometheus/prometheus.go (hunk 5 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/metrics/prometheus/prometheus.go:266]: we need it for dashboard

File: modules/pubmatic/openwrap/beforevalidationhook.go [hunk 12]
```diff
@@ -669,6 +675,20 @@ func (m *OpenWrap) applyBannerAdUnitConfig(rCtx models.RequestCtx, imp *openrtb2
 	}
 }
 
+// isVastUnwrapEnabled return whether to enable vastunwrap or not
+func isVastUnwrapEnabled(PartnerConfigMap map[int]map[string]string, VASTUnwrapTraffic int) bool {
+	trafficPercentage := VASTUnwrapTraffic
+	unwrapEnabled := models.GetVersionLevelPropertyFromPartnerConfig(PartnerConfigMap, models.VastUnwrapperEnableKey) == VastUnwrapperEnableValue
+	if unwrapEnabled {
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/beforevalidationhook.go:682]: There are few profiles without trafficPercentDB, in which we need to use unwrapEnabled flag from DB and trafficPercentage from config file.

File: modules/pubmatic/openwrap/unwrap_service.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+			m.metricEngine.RecordUnwrapRespTime(accountID, strconv.Itoa(int(wrapperCnt)), respTime)
+		}
+	}()
+	headers := http.Header{}
+	headers.Add(models.ContentType, "application/xml; charset=utf-8")
+	headers.Add(models.UserAgent, userAgent)
+	headers.Add(models.XUserAgent, userAgent)
+	headers.Add(models.XUserIP, ip)
+	headers.Add(models.CreativeID, bid.Bid.ID)
+	headers.Add(models.UnwrapTimeout, strconv.Itoa(m.cfg.VastUnwrapCfg.APPConfig.UnwrapDefaultTimeout))
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/unwrap_service.go]: Done

File: modules/pubmatic/openwrap/unwrap_service.go [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/modules/pubmatic/openwrap/models"
+)
+
+func (m OpenWrap) doUnwrapandUpdateBid(isStatsEnabled bool, bid *adapters.TypedBid, userAgent string, ip string, unwrapURL string, accountID string, bidder string) {
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/unwrap_service.go]: Done

File: modules/pubmatic/openwrap/config/config.go [hunk 15]
```diff
@@ -80,6 +82,8 @@ type PixelView struct {
 }
 
 type FeatureToggle struct {
+	VASTUnwrapPecent      int
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/config/config.go:85]: Done

File: modules/pubmatic/openwrap/unwrap_service.go (hunk 6 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/unwrap_service.go]: moved code into unwrap package

**Review by Pubmatic-Dhruv-Sonone — DISMISSED** [review]

File: modules/pubmatic/openwrap/abtest.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/abtest.go:25]: Looks like we are still copying the map.

File: modules/pubmatic/openwrap/beforevalidationhook.go [hunk 16]
```diff
@@ -153,6 +153,12 @@ func (m OpenWrap) handleBeforeValidationHook(
 		result.Warnings = append(result.Warnings, "update the rCtx.PartnerConfigMap with ABTest data")
 	}
 
+	// To check if VAST unwrap needs to be enabled for given request
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/beforevalidationhook.go:156]: Add UTs for this

File: modules/pubmatic/openwrap/openwrap.go [hunk 17]
```diff
@@ -78,13 +81,18 @@ func initOpenWrap(rawCfg json.RawMessage, moduleDeps moduledeps.ModuleDeps) (Ope
...
+	vastunwrap.InitUnWrapperConfig(cfg.VastUnwrapCfg)
+
 	once.Do(func() {
 		ow = &OpenWrap{
 			cfg:                cfg,
 			cache:              owCache,
 			metricEngine:       &metricEngine,
 			currencyConversion: moduleDeps.CurrencyConversion,
 			featureConfig:      featureConfig,
+			unwrap: unwrap.NewUnwrap(fmt.Sprintf("http://%s:%d/unwrap", cfg.VastUnwrapCfg.APPConfig.Host, cfg.VastUnwrapCfg.APPConfig.Port),
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/openwrap.go:94]: Can we move this outside and set it in var and use it here? Would look cleaner.

File: modules/pubmatic/openwrap/openwrap_sshb.go [hunk 18]
```diff
@@ -52,27 +49,7 @@ func (ow *OpenWrap) GetFeature() publisherfeature.Feature {
...
-		if value := models.GetVersionLevelPropertyFromPartnerConfig(rCtx.PartnerConfigMap, models.VastUnwrapTrafficPercentKey); len(value) > 0 {
-			if trafficPercentDB, err := strconv.Atoi(value); err == nil {
-				trafficPercentage = trafficPercentDB
-			}
-		}
-	}
-	return unwrapEnabled && GetRandomNumberIn1To100() <= trafficPercentage
+func getVastUnwrapperEnable(ctx context.Context, field string) bool {
+	vastEnableUnwrapper, _ := ctx.Value(field).(string)
+	return vastEnableUnwrapper == "1"
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/openwrap_sshb.go:54]: Use constant

File: modules/pubmatic/openwrap/unwrap/unwrap.go [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+	httpResp := NewCustomRecorder()
+	uw.unwrapRequest(httpResp, httpReq)
+	respStatus = httpResp.Header().Get(models.UnwrapStatus)
+	wrapperCnt, _ = strconv.ParseInt(httpResp.Header().Get(models.UnwrapCount), 10, 0)
+	if !isStatsEnabled && httpResp.Code == http.StatusOK && respStatus == "0" {
+		respBody := httpResp.Body.Bytes()
+		bid.Bid.AdM = string(respBody)
+	}
+
+	glog.V(3).Infof("[VAST_UNWRAPPER] pubid:[%v] bidder:[%v] impid:[%v] bidid:[%v] status_code:[%v] wrapper_cnt:[%v] httpRespCode= [%v] statsEnabled:[%v]",
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/unwrap/unwrap.go:86]: Can we make this constant and create a generic  logging template?

File: modules/pubmatic/openwrap/beforevalidationhook.go (hunk 16 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/beforevalidationhook.go:156]: Already added UT TestIsVastUnwrapEnabled

File: modules/pubmatic/openwrap/beforevalidationhook.go [hunk 20]
```diff
@@ -669,6 +675,20 @@ func (m *OpenWrap) applyBannerAdUnitConfig(rCtx models.RequestCtx, imp *openrtb2
 	}
 }
 
+// isVastUnwrapEnabled return whether to enable vastunwrap or not
+func isVastUnwrapEnabled(PartnerConfigMap map[int]map[string]string, VASTUnwrapTraffic int) bool {
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/beforevalidationhook.go:679]: Done

File: modules/pubmatic/openwrap/openwrap.go (hunk 17 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/openwrap.go:94]: Done

File: modules/pubmatic/openwrap/openwrap_sshb.go (hunk 18 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/openwrap_sshb.go:54]: Done

File: modules/pubmatic/openwrap/unwrap/unwrap.go (hunk 19 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/unwrap/unwrap.go:86]: Adding generic template would not be good idea as in each print log fields will be changed

File: modules/pubmatic/openwrap/abtest.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/abtest.go:25]: We are not using this function, removed it

**Review by Pubmatic-Supriya-Patil — APPROVED** [review]

File: modules/pubmatic/openwrap/beforevalidationhook.go (hunk 16 above)
Comment (PubMatic-OpenWrap) [reply modules/pubmatic/openwrap/beforevalidationhook.go:156]: TestIsVastUnwrapEnabled covers only the function. We should add UT to address line 157 to 160 (UT of handleBeforeValidationHook )

File: modules/pubmatic/openwrap/unwrap/unwrap.go [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+	"time"
+
+	vastunwrap "git.pubmatic.com/vastunwrap"
+	"github.com/golang/glog"
+	"github.com/prebid/prebid-server/adapters"
+	metrics "github.com/prebid/prebid-server/modules/pubmatic/openwrap/metrics"
+	"github.com/prebid/prebid-server/modules/pubmatic/openwrap/models"
+)
+
+type Unwrap struct {
```
Comment (PubMatic-OpenWrap) [inline modules/pubmatic/openwrap/unwrap/unwrap.go:17]: Ideally Unwrap type should not be exposed if `NewUnwrap()` needs to be enforced

**Review by PubMatic-OpenWrap — APPROVED** [review]
Approved with minor comment.
Please address Dhruv's comments

File: modules/pubmatic/openwrap/unwrap/unwrap.go (hunk 21 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/unwrap/unwrap.go:17]: Unwrap type is exposed to add in OpenWrap object which would be initialised during init()

File: modules/pubmatic/openwrap/beforevalidationhook.go (hunk 16 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/beforevalidationhook.go:156]: Added UT's

**Review by Pubmatic-Dhruv-Sonone — APPROVED** [review]

**Review by Pubmatic-Supriya-Patil — APPROVED** [review]

**Pull request merged**